### PR TITLE
Follow pagination and implement rate throttling

### PR DESCRIPTION
**Pull request by systemist**
I was having a problem with images not getting deleted. It turns out the
maximum page size is now 100, down from 500. Since the application I'm
using this for consistently has more than 100 images that don't meet the
delete criteria and they're in the first page, the action would never
fetch the images that needed to be deleted.

This uses the Hypermedia `link` header
(https://docs.github.com/en/rest/overview/resources-in-the-rest-api#hypermedia)
to traverse all pages. This worked well, but I hit both primary and
secondary rate limits trying to delete all my old images, so I also
added a function to handle rate limiting.

This still needs tests, documentation, and probably some error handling
and print statements, but I wanted to open a PR to see whether you're
okay with this approach and would consider merging it before I spend on
these details.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -135,9 +152,34 @@ async def list_package_versions(*, image_name: ImageName, http_client: AsyncClie
...
-    return [PackageVersionResponse(**i) for i in response.json()]
+    packages = await get_all_pages(
+        url=f'{BASE_URL}/user/packages/container/{image_name.encoded}/versions?per_page=100', http_client=http_client
+    )
+    return [PackageVersionResponse(**i) for i in packages]
+
+
+async def get_all_pages(*, url: str, http_client: AsyncClient) -> list[dict]:
+    """
+    Accumulate all pages of an paginated API endpoint
```
Comment (sondrelg) [inline main.py:163]:
```suggestion
    Accumulate all pages of a paginated API endpoint.
```

**Review by sondrelg — COMMENTED** [review]
Traversing backwards and handling rate limits both sound like great ideas, and the current implementation looks good to me at first glance 👏  Let me know when you've added docs and tests, and I'll review right away 👍

**Review by lukasz-mitka — COMMENTED** [review]
Please add pagination for `list_packages` and `list_org_packages`. Same issue is present there.
https://github.com/snok/container-retention-policy/pull/42/files#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R95

**Comment by sondrelg** [comment]
Any updates here @systemist? 🙂

**Comment by den-is** [comment]
I've encountered the same issue
One ugly configured repository has 10k untagged and 2k tagged images.
This update and work will be highly appreciated! 🙇

**Comment by sondrelg** [comment]
@systemist are you still planning on finishing this?

**Comment by sondrelg** [comment]
Since @systemist seems like he won't finish this. Does anyone else fancy finishing the PR?

**Comment by sondrelg** [comment]
I'll merge this, try to add tests and docs, then upgrade today hopefully :slightly_smiling_face:

**Comment by sondrelg** [comment]
The main branch should now be up to date. I've added some tests, and I've added pagination support to the package fetching.

Does anyone have a repo with a package count high enough to test if the main branch currently works? Would be nice to get at least one confirmation before we bump major version, if anyone here is free to try :bow:

**Comment by lukasz-mitka** [comment]
@sondrelg Thank you for finishing this!
It's working fine in my setup (>>100 versions)

Found the bug though. Step outputs didn't get set, could you look into that?

**Comment by sondrelg** [comment]
Thanks for testing! I'll release a new version today then hopefully 👍

**Pull request merged**